### PR TITLE
Polars: boolean binary operations

### PR DESCRIPTION
**Pull request by Shoeboxam**

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1600?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1597** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1597?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1596** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1596?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1592** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1592?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1591** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1591?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1534** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1600** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1600?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1539** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1540** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1586** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1590** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1590?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1584** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1584?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1542** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1542?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1503](https://github.com/opendp/opendp/pull/1503) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1503?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1506](https://github.com/opendp/opendp/pull/1506) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1566** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/transformations/make_stable_expr/expr_binary/mod.rs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        frame_domain,
+        context,
+    } = input_domain.clone();
+
+    let expr_domain = ExprDomain::new(frame_domain, ExprContext::RowByRow);
+    let t_left = left.make_stable(expr_domain.clone(), input_metric.clone())?;
+    let t_right = right.make_stable(expr_domain.clone(), input_metric.clone())?;
+
+    use polars_plan::dsl::Operator::*;
+    if !matches!(op, Eq | NotEq | Lt | LtEq | Gt | GtEq | And | Or | Xor) {
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_binary/mod.rs:48]: I've found https://docs.rs/polars-lazy/latest/polars_lazy/dsl/enum.Operator.html, but I haven't found definitions for these. In particular, what about `LogicalAnd` and `LogicalOr`? Are they just aliases? Or are plain `And` and `Or` bitwise?

File: rust/src/transformations/make_stable_expr/expr_binary/mod.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        return fallible!(MakeTransformation, "unsupported operator: {:?}. Only binary operations that emit booleans are currently supported.", op);
+    }
+
+    let mut series_domain =
+        SeriesDomain::new(&*expr_output_name(&expr)?, AtomDomain::<bool>::default());
+
+    series_domain.nullable = t_left.output_domain.active_series()?.nullable
+        || t_right.output_domain.active_series()?.nullable;
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_binary/mod.rs:56]: Does polars define the semantics of these operations? I'm wondering if for the boolean operations `&&` might be appropriate here. For example I'd expect `TRUE || NULL == TRUE` rather than `TRUE || NULL == NULL`, but maybe that's not the way it works.

File: rust/src/transformations/make_stable_expr/expr_binary/test.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    };
+}
+
+#[test]
+fn test_binary_ops() -> Fallible<()> {
+    test_binary!(
+        get_f64_data,
+        eq,
+        [
+            Some(true),
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_binary/test.rs:75]: Repeating a stylistic preference from a different PR, but I think it would be easier to understand the tests if the things we were comparing were next to each other, even if it meant the tests were longer. The comments in the assertions below are good, but you need to go from the bottom of the file to the top to see where the data is coming from.

I'd also prefer that to have these as separate tests, so if something does go wrong, we see all the errors, instead of only the first one.

... but these are pretty minor concerns, and probably not worth the refactoring required, even if I've been persuasive.

**Review by mccalluc — COMMENTED** [review]
A couple questions about how you arrived at this set of operators and the semantics if we do have a null. If everything is in fact working the way it should, consider adding the answers to the questions as comments in the code, so they are there for the future.

File: rust/src/transformations/make_stable_expr/expr_binary/mod.rs (hunk 2 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_binary/mod.rs:56]: Based on the tests, `and` and `xor` both propagate nulls. Therefore all operations are sound, and all operations except for `or` are complete. I added conditions here to make it complete as well.

Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_binary/mod.rs:56]: After more testing, lhs nullity or falsey can still propagate null. I'm just leaving it simple, if either side can be null, the output can be null.

File: rust/src/transformations/make_stable_expr/expr_binary/test.rs (hunk 3 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_binary/test.rs:75]: I made some changes, and also print the input L R dataframe when something fails, so you can at least compare in the failure output.

**Review by mccalluc — APPROVED** [review]
